iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp

```diff
@@ -0,0 +1,307 @@
+    if (auto alloc = dyn_cast<AllocOp>(op)) analyzeAlloc(alloc);
+  });
+}
+
+void MemRefUsageAnalysis::analyzeFunc(FuncOp funcOp) {
+  for (Value arg : funcOp.getArguments()) {
+    SmallVector<Operation *, 4> vectorUses;
+    if (memRefCanBeVectorized(arg.getType()) &&
+        allVectorUses(arg, vectorUses)) {
+      vectorize.insert(arg);
+      transferOps.insert(vectorUses.begin(), vectorUses.end());
+    }
+  }
+}
+
+bool MemRefUsageAnalysis::allVectorUses(Value v,
+                                        SmallVector<Operation *, 4> &uses) {
+  for (Operation *userOp : v.getUsers()) {
+    // Ignore dealloc ops as they translate to noop.
+    if (isa<DeallocOp>(userOp)) continue;
+    // Only vectorize memref used by vector transfer ops.
+    if (!isa<vector::TransferReadOp, vector::TransferWriteOp>(userOp))
+      return false;
+    uses.push_back(userOp);
+  }
+  return true;
+}
+
+void MemRefUsageAnalysis::analyzeAlloc(AllocOp allocOp) {
+  SmallVector<Operation *, 4> vectorUses;
+  if (memRefCanBeVectorized(allocOp.getType()) &&
+      allVectorUses(allocOp, vectorUses)) {
+    vectorize.insert(allocOp);
+    transferOps.insert(vectorUses.begin(), vectorUses.end());
+  }
+}
+
+template <typename OpTy>
+class MemRefConversionPattern : public OpConversionPattern<OpTy> {
+ public:
+  MemRefConversionPattern<OpTy>(MLIRContext *context,
+                                const MemRefUsageAnalysis &memrefUsageAnalysis)
+      : OpConversionPattern<OpTy>::OpConversionPattern(context),
+        memrefUsageAnalysis(memrefUsageAnalysis) {}
+
+ protected:
+  const MemRefUsageAnalysis &memrefUsageAnalysis;
+};
+
+class ProcessFuncArg final : public MemRefConversionPattern<FuncOp> {
+ public:
+  using MemRefConversionPattern<FuncOp>::MemRefConversionPattern;
+  LogicalResult matchAndRewrite(
+      FuncOp funcOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override;
+};
+
+class ProcessTransferRead final
+    : public MemRefConversionPattern<vector::TransferReadOp> {
+ public:
+  using MemRefConversionPattern<
+      vector::TransferReadOp>::MemRefConversionPattern;
+  LogicalResult matchAndRewrite(
+      vector::TransferReadOp read, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    if (!memrefUsageAnalysis.transferConvert(read)) return failure();
+    vector::TransferReadOp::Adaptor adaptor(operands);
+    Value memref = adaptor.memref();
+    Location loc = read.getLoc();
+    unsigned ratio =
+        getBitWidth(memref.getType().cast<MemRefType>().getElementType()) /
+        getBitWidth(read.getMemRefType().getElementType());
+    SmallVector<Value, 4> indices(adaptor.indices().begin(),
+                                  adaptor.indices().end());
+    indices.back() = rewriter.create<SignedDivIOp>(
+        loc, indices.back(), rewriter.create<ConstantIndexOp>(loc, ratio));
+    if (getBitWidth(memref.getType().cast<MemRefType>().getElementType()) ==
```

Comment:
I am not sure what is being handled in the `if` and the `else` parts. If the bitwidths are the same already, I would figure its a simpler replacement.

Reply:
Looking at the example below, not clear to me we really want to be replacing with the `load` directly. Could we just use a new `vector.transfer_read` like the `else` part does?

Reply:
Isnt this a missing lowering in vector.transfer_read? If the memref element type is a vector, then you could emit a vector_load/vector_store over there. This seems like it is being down prematurely.

Reply:
Why is this premature to do it here? We could emit a transfer_read strictly equivalent to the load/store and lower it to a spirv Load/Store, but we still need to emit the bitcast/shape_cast so that they can get optimized away before converting to spirv. Isn't it simpler to always try to lower std load/store before getting to spirv conversion? Or are you suggesting adding an extra pass?